loguru.hpp

```diff
@@ -1379,29 +1392,100 @@ namespace loguru
 	const char* terminal_reset()      { return s_terminal_has_color ? "\e[0m" : ""; }
 
 	// ------------------------------------------------------------------------------
-
+#ifdef LOGURU_WITH_FILEABS
+	void file_reopen(void* user_data);
+	inline FILE* to_file(void* user_data) { return reinterpret_cast<FileAbs*>(user_data)->fp; }
+#else
+	inline FILE* to_file(void* user_data) { return reinterpret_cast<FILE*>(user_data); }
+#endif
 	void file_log(void* user_data, const Message& message)
 	{
-		FILE* file = reinterpret_cast<FILE*>(user_data);
+#ifdef LOGURU_WITH_FILEABS
+		pthread_mutex_lock(&(reinterpret_cast<FileAbs*>(user_data)->lock));
+		file_reopen(user_data);
+		FILE* file = to_file(user_data);
+		if(!file) {
+			pthread_mutex_unlock(&(reinterpret_cast<FileAbs*>(user_data)->lock));
+			return;
+		}
+#else
+		FILE* file = to_file(user_data);
+#endif
 		fprintf(file, "%s%s%s%s\n",
 			message.preamble, message.indentation, message.prefix, message.message);
 		if (g_flush_interval_ms == 0) {
 			fflush(file);
 		}
+#ifdef LOGURU_WITH_FILEABS
+		pthread_mutex_unlock(&(reinterpret_cast<FileAbs*>(user_data)->lock));
+#endif
 	}
 
 	void file_close(void* user_data)
 	{
-		FILE* file = reinterpret_cast<FILE*>(user_data);
-		fclose(file);
+		FILE* file = to_file(user_data);
+		if(file) {
+			fclose(file);
+		}
+#ifdef LOGURU_WITH_FILEABS
+		delete reinterpret_cast<FileAbs*>(user_data);
+#endif
 	}
 
 	void file_flush(void* user_data)
 	{
-		FILE* file = reinterpret_cast<FILE*>(user_data);
+		FILE* file = to_file(user_data);
 		fflush(file);
 	}
 
+#ifdef LOGURU_WITH_FILEABS
+	void add_callback_abs(const char* id, log_handler_t callback, void* user_data,
+						  Verbosity verbosity, close_handler_t on_close, flush_handler_t on_flush)
+	{
+		std::lock_guard<std::recursive_mutex> lock(s_mutex);
+		s_callbacks.push_back(Callback{id, callback, user_data, verbosity, on_close, on_flush, 0});
+	}
+
+	bool remove_callback_abs(const char* id)
+	{
+		std::lock_guard<std::recursive_mutex> lock(s_mutex);
+		auto it = std::find_if(begin(s_callbacks), end(s_callbacks), [&](const Callback& c) { return c.id == id; });
+		if (it != s_callbacks.end()) {
+			s_callbacks.erase(it);
+			return true;
+		} else {
+			LOG_F(ERROR, "Failed to locate callback with id '%s'", id);
+			return false;
+		}
+	}
+
+	void file_reopen(void* user_data)
+	{
+		FileAbs * file_abs = reinterpret_cast<FileAbs*>(user_data);
+		struct stat st;
+		int ret;
+		if (!file_abs->fp || (ret = stat(file_abs->path, &st)) == -1 || (st.st_ino != file_abs->st.st_ino)) {
+			remove_callback_abs(file_abs->path); //adapted version, only remove from callback vector s_callbacks;
+			if (ret < 0) {
+				LOG_F(INFO, "Reopening file '%s' due to %m", file_abs->path); // still logging to previous fd;
```

Comment:
This is the first time I see `%m` - is that supported by Clang? I'd prefer more portable code - in particular, this is cross-platform (and thread-safe):

```
const auto why = errno_as_text();
LOG_F(INFO, "Reopening file '%s' due to %s", file_abs->path, why.c_str());
```